core/src/kura.rs

```diff
@@ -4,258 +4,504 @@
+            block_hash_array: Mutex::new(Vec::new()),
             wsv,
             broker,
-            actor_channel_capacity,
-        })
+            block_reciever: Mutex::new(block_reciever),
+            block_sender,
+        });
+
+        // How do we avoid the spawned thread taking on a
+        // life of it's own, aka keeping kura alive and running
+        // long after it has been dropped by all who needed it.
+        // By using a weak reference, see `kura_recieve_blocks_thread`
+        // for details on how this works.
+        let kura_weak_reference = Arc::downgrade(&kura);
+        std::thread::spawn(move || {
+            Self::kura_recieve_blocks_thread(&kura_weak_reference);
+        });
+        Ok(kura)
     }
-}
-
-/// Generic kura trait for mocks
-#[async_trait]
-pub trait KuraTrait:
-    Actor
-    + ContextHandler<StoreBlock, Result = ()>
-    + ContextHandler<GetBlockHash, Result = Option<HashOf<VersionedCommittedBlock>>>
-    + Debug
-{
-    /// Construct [`Kura`].
-    /// Kura will not be ready to work with before `init()` method invocation.
-    /// # Errors
-    /// Fails if reading from disk while initing fails
-    async fn new(
-        mode: Mode,
-        block_store_path: &Path,
-        blocks_per_file: NonZeroU64,
-        wsv: Arc<WorldStateView>,
-        broker: Broker,
-        actor_channel_capacity: u32,
-    ) -> Result<Self>;
 
     /// Loads kura from configuration
     /// # Errors
-    /// Fails if call to new fails
-    async fn from_configuration(
+    /// Fails if there are filesystem errors when trying
+    /// to access the block store indicated by the
+    /// path in the configuration.
+    pub fn from_configuration(
         configuration: &config::KuraConfiguration,
         wsv: Arc<WorldStateView>,
         broker: Broker,
-    ) -> Result<Self> {
+    ) -> Result<Arc<Self>> {
         Self::new(
             configuration.init_mode,
             Path::new(&configuration.block_store_path),
-            configuration.blocks_per_storage_file,
             wsv,
             broker,
             configuration.actor_channel_capacity,
         )
-        .await
     }
-}
 
-#[async_trait]
-impl KuraTrait for Kura {
-    async fn new(
-        mode: Mode,
-        block_store_path: &Path,
-        blocks_per_file: NonZeroU64,
-        wsv: Arc<WorldStateView>,
-        broker: Broker,
-        mailbox: u32,
-    ) -> Result<Self> {
-        Self::new_meta(
-            mode,
-            block_store_path,
-            blocks_per_file,
-            wsv,
-            broker,
-            mailbox,
-            DefaultIO,
-        )
-        .await
+    /// Initialize [`Kura`] after its construction to be able to work with it.
+    ///
+    /// # Errors
+    /// Fails if:
+    /// - file storage is unavailable
+    /// - data in file storage is invalid or corrupted
+    #[allow(clippy::unwrap_in_result, clippy::expect_used)]
+    pub fn init(&self) -> Result<Vec<VersionedCommittedBlock>> {
```

Comment:
It's split up this way because we want the option of re-initializing kura. We may decide down the line that we don't ever need to do that and could then merge it with new. This aspect is inherited from the way kura was previously.